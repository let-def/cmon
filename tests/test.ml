let print name value =
  print_string name;
  print_endline ":";
  PPrint.ToChannel.pretty 0.9 80 stdout (Cmon.print value);
  print_newline ();
  print_newline ()

let short_args = Cmon.[int 1; bool true]
let mid_args = short_args @ short_args @ short_args
let long_args = mid_args @ mid_args @ mid_args

let short_fields = Cmon.["a", int 1; "b", bool true]
let mid_fields = short_fields @ short_fields @ short_fields
let long_fields = mid_fields @ mid_fields @ mid_fields

let () =
  (* Base types *)
  print "unit"  Cmon.unit;
  print "false" (Cmon.bool false);
  print "true"  (Cmon.bool true);

  (* Integers *)
  print "0"       (Cmon.int 0);
  print "1"       (Cmon.int 1);
  print "2"       (Cmon.int 2);
  print "-1"      (Cmon.int (-1));
  print "-2"      (Cmon.int (-2));
  print "max_int" (Cmon.int max_int);
  print "min_int" (Cmon.int min_int);

  (* Floats *)
  print "0."        (Cmon.float 0.0);
  print "1."        (Cmon.float 1.0);
  print "-0."       (Cmon.float (-0.0));
  print "-1."       (Cmon.float (-1.0));
  print "+inf"      (Cmon.float ( 1.0 /. 0.0));
  print "-inf"      (Cmon.float (-1.0 /. 0.0));
  print "nan"       (Cmon.float ( 0.0 /. 0.0));
  print "max_float" (Cmon.float max_float);
  print "min_float" (Cmon.float min_float);

  (* Chars *)
  print "'a'"    (Cmon.char 'a');
  print "'\\x00'" (Cmon.char '\x00');
  print "'\\xFF'" (Cmon.char '\xFF');
  print "'\\t'"    (Cmon.char '\t');
  print "'\\n'"    (Cmon.char '\n');

  (* Strings *)
  print "\"foo\"" (Cmon.string "foo");
  print "\"\\\"foo\\\"\"" (Cmon.string "\"foo\"");
  print "\"foo\\nbar\"" (Cmon.string "foo\nbar");

  (* Tuple *)
  print "(1, true)" (Cmon.tuple short_args);
  print "(1, true, ...)" (Cmon.tuple mid_args);
  print "(1, true, ..., ...)" (Cmon.tuple long_args);

  (* Terminated lists *)
  print "[1; true]" (Cmon.list short_args);
  print "[1; true; ...]" (Cmon.list mid_args);
  print "[1; true; ...; ...]" (Cmon.list long_args);

  (* Open lists *)
  let open_list xs = List.fold_right Cmon.cons xs (Cmon.constant "xs") in
  print "1 :: true :: xs" (open_list short_args);
  print "1 :: true :: ... :: xs" (open_list mid_args);
  print "1 :: true :: ... ... :: xs" (open_list long_args);

  (* Nested lists and tuples *)
  let step = ref 0 in
  let rec deep n =
    incr step;
    let is_tuple = !step land 1 = 0 in
    if n = 0 then (
      if is_tuple
      then Cmon.unit
      else Cmon.nil
    ) else (
      let left = deep (n - 1) in
      let right = deep (n - 1) in
      if is_tuple
      then Cmon.tuple [left; right]
      else Cmon.list [left; right]
    )
  in
  print "nested-1" (deep 3);
  print "nested-2" (deep 6);

  (* Constructors *)
  print "data-constructors-1"
    (Cmon.tuple [
        Cmon.constant "None";
        Cmon.constructor "Some" Cmon.unit;
        Cmon.constructor "Pair" (Cmon.tuple [Cmon.int 1; Cmon.int 2]);
      ]);

  print "data-constructors-2"
    (Cmon.tuple [
        Cmon.construct "None" [];
        Cmon.construct "Some" [Cmon.unit];
        Cmon.construct "Pair" [Cmon.int 1; Cmon.int 2];
      ]);

  (* Records *)
  print "record-1" (Cmon.record ["contents", Cmon.int 1]);
  print "record-2" (Cmon.record short_fields);
  print "record-3" (Cmon.record mid_fields);
  print "record-4" (Cmon.record long_fields);

  (* Inline records *)
  print "inline-record-1" (Cmon.crecord "R1" ["contents", Cmon.int 1]);
  print "inline-record-2" (Cmon.crecord "R2" short_fields);
  print "inline-record-3" (Cmon.crecord "R3" mid_fields);
  print "inline-record-4" (Cmon.crecord "R4" long_fields);

  (* Sharing test *)
  let shared_term1 =
    Cmon.tuple [Cmon.string "test"; Cmon.int 1] in
  let shared_term2 =
    Cmon.tuple [shared_term1; shared_term1] in
  let shared_term1' =
    Cmon.tuple [Cmon.string "test"; Cmon.int 1] in
  let shared_term2' =
    Cmon.tuple [shared_term1'; shared_term1'] in
  let shared_term3 =
    Cmon.tuple [shared_term2; shared_term2; shared_term2'] in
  let shared_term4 =
    Cmon.tuple [shared_term3; shared_term2'] in
  print "shared-terms" shared_term4
