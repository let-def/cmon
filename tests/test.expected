unit:
()

false:
false

true:
true

0:
0

1:
1

2:
2

-1:
-1

-2:
-2

max_int:
4611686018427387903

min_int:
-4611686018427387904

0.:
0.

1.:
1.

-0.:
-0.

-1.:
-1.

+inf:
infinity

-inf:
neg_infinity

nan:
nan

max_float:
1.79769313486231571e+308

min_float:
2.22507385850720138e-308

'a':
'a'

'\x00':
'\000'

'\xFF':
'\255'

'\t':
'\t'

'\n':
'\n'

"foo":
"foo"

"\"foo\"":
"\"foo\""

"foo\nbar":
"foo\nbar"

(1, true):
(1, true)

(1, true, ...):
(1, true, 1, true, 1, true)

(1, true, ..., ...):
(
  1,
  true,
  1,
  true,
  1,
  true,
  1,
  true,
  1,
  true,
  1,
  true,
  1,
  true,
  1,
  true,
  1,
  true
)

[1; true]:
[ 1; true ]

[1; true; ...]:
[ 1; true; 1; true; 1; true ]

[1; true; ...; ...]:
[
  1;
  true;
  1;
  true;
  1;
  true;
  1;
  true;
  1;
  true;
  1;
  true;
  1;
  true;
  1;
  true;
  1;
  true
]

1 :: true :: xs:
1 :: true :: xs

1 :: true :: ... :: xs:
1 :: true :: 1 :: true :: 1 :: true :: xs

1 :: true :: ... ... :: xs:
1
:: true
:: 1
:: true
:: 1
:: true
:: 1
:: true
:: 1
:: true
:: 1
:: true
:: 1
:: true
:: 1
:: true
:: 1
:: true
:: xs

nested-1:
[ ([ (); [] ], ([], ())); [ ([], ()); [ (); [] ] ] ]

nested-2:
(
  [
    (
      [ ([ (); [] ], ([], ())); [ ([], ()); [ (); [] ] ] ],
      ([ ([], ()); [ (); [] ] ], ([ (); [] ], ([], ())))
    );
    [
      ([ ([], ()); [ (); [] ] ], ([ (); [] ], ([], ())));
      [ ([ (); [] ], ([], ())); [ ([], ()); [ (); [] ] ] ]
    ]
  ],
  (
    [
      ([ ([], ()); [ (); [] ] ], ([ (); [] ], ([], ())));
      [ ([ (); [] ], ([], ())); [ ([], ()); [ (); [] ] ] ]
    ],
    (
      [ ([ (); [] ], ([], ())); [ ([], ()); [ (); [] ] ] ],
      ([ ([], ()); [ (); [] ] ], ([ (); [] ], ([], ())))
    )
  )
)

data-constructors-1:
(None, Some (), Pair (1, 2))

data-constructors-2:
(None, Some (), Pair (1, 2))

record-1:
{ contents = 1; }

record-2:
{ a = 1; b = true; }

record-3:
{ a = 1; b = true; a = 1; b = true; a = 1; b = true; }

record-4:
{
  a = 1;
  b = true;
  a = 1;
  b = true;
  a = 1;
  b = true;
  a = 1;
  b = true;
  a = 1;
  b = true;
  a = 1;
  b = true;
  a = 1;
  b = true;
  a = 1;
  b = true;
  a = 1;
  b = true;
}

inline-record-1:
R1 { contents = 1; }

inline-record-2:
R2 { a = 1; b = true; }

inline-record-3:
R3 { a = 1; b = true; a = 1; b = true; a = 1; b = true; }

inline-record-4:
R4 {
  a = 1;
  b = true;
  a = 1;
  b = true;
  a = 1;
  b = true;
  a = 1;
  b = true;
  a = 1;
  b = true;
  a = 1;
  b = true;
  a = 1;
  b = true;
  a = 1;
  b = true;
  a = 1;
  b = true;
}

shared-terms:
let v0 = let v1 = ("test", 1) in (v1, v1) in
(let v2 = let v3 = ("test", 1) in (v3, v3) in (v2, v2, v0), v0)

rec-term1:
let rec v0 = 1 :: v0 in
v0

rec-term2:
let rec v0 = 1 :: 2 :: v0 in
v0

rec-term4:
let rec v0 = 1 :: 2 :: v0 in
v0

rec-term4':
let rec v1 = 2 :: v0 and v0 = 1 :: v1 in
(v0, v1)

rec-term5:
let rec v2 = 2 :: v1 and v1 = 1 :: v2 in
let v0 = ("test", 1) in
(v1, v2, v0, v0)

rec-term6:
let rec v2 = ("test", 1) and v1 = v2 :: v0 and v0 = v2 :: v1 in
(v0, v1)

rec-term7:
let rec v2 = ("test", 1) and v1 = v2 :: v0 and v0 = v2 :: v1 in
(v0, v1, (v2, v2))

